go/common/crypto/signature/signers/remote/grpc.go

```diff
@@ -0,0 +1,277 @@
+
+	methodPublicKeys = serviceName.NewMethod("PublicKeys", nil)
+	methodSign       = serviceName.NewMethod("Sign", SignRequest{})
+
+	serviceDesc = grpc.ServiceDesc{
+		ServiceName: string(serviceName),
+		HandlerType: (*wrapper)(nil),
+		Methods: []grpc.MethodDesc{
+			{
+				MethodName: methodPublicKeys.ShortName(),
+				Handler:    handlerPublicKeys,
+			},
+			{
+				MethodName: methodSign.ShortName(),
+				Handler:    handlerSign,
+			},
+		},
+	}
+)
+
+// PublicKey is a public key supported by the remote signer.
+type PublicKey struct {
+	Role      signature.SignerRole `json:"role"`
+	PublicKey signature.PublicKey  `json:"public_key"`
+}
+
+// SignRequest is a signature request.
+type SignRequest struct {
+	Role    signature.SignerRole `json:"role"`
+	Context string               `json:"context"`
+	Message []byte               `json:"message"`
+}
+
+type wrapper struct {
+	signers map[signature.SignerRole]signature.Signer
+}
+
+func (w *wrapper) publicKeys(ctx context.Context) ([]PublicKey, error) {
+	var resp []PublicKey
+	for _, v := range signature.SignerRoles { // Return in consistent order.
+		if signer := w.signers[v]; signer != nil {
+			resp = append(resp, PublicKey{
+				Role:      v,
+				PublicKey: signer.Public(),
+			})
+		}
+	}
+	return resp, nil
+}
+
+func (w *wrapper) sign(ctx context.Context, req *SignRequest) ([]byte, error) {
+	signer, ok := w.signers[req.Role]
+	if !ok {
+		return nil, signature.ErrNotExist
+	}
+	return signer.ContextSign(signature.Context(req.Context), req.Message)
+}
+
+func handlerPublicKeys( // nolint: golint
+	srv interface{},
+	ctx context.Context,
+	dec func(interface{}) error,
+	interceptor grpc.UnaryServerInterceptor,
+) (interface{}, error) {
+	if interceptor == nil {
+		return srv.(*wrapper).publicKeys(ctx)
+	}
+	info := &grpc.UnaryServerInfo{
+		Server:     srv,
+		FullMethod: methodPublicKeys.FullName(),
+	}
+	handler := func(ctx context.Context, req interface{}) (interface{}, error) {
+		return srv.(*wrapper).publicKeys(ctx)
+	}
+	return interceptor(ctx, nil, info, handler)
+}
+
+func handlerSign( // nolint: golint
+	srv interface{},
+	ctx context.Context,
+	dec func(interface{}) error,
+	interceptor grpc.UnaryServerInterceptor,
+) (interface{}, error) {
+	var req SignRequest
+	if err := dec(&req); err != nil {
+		return nil, err
+	}
+	if interceptor == nil {
+		return srv.(*wrapper).sign(ctx, &req)
+	}
+	info := &grpc.UnaryServerInfo{
+		Server:     srv,
+		FullMethod: methodSign.FullName(),
+	}
+	handler := func(ctx context.Context, req interface{}) (interface{}, error) {
+		return srv.(*wrapper).sign(ctx, req.(*SignRequest))
+	}
+	return interceptor(ctx, &req, info, handler)
+}
+
+// RegisterService registers a new remote signer backend service with the given
+// gRPC server.
+//
+// WARNING: NEVER call this from the actual node.
+func RegisterService(server *grpc.Server, signerFactory signature.SignerFactory) {
+	// Not sure if this is the best place to do this.
```

Comment:
Probably not, I would rather do this in the actual remote signer binary entry point?